### PR TITLE
Typescript Types Support (Ready for Review)

### PR DESCRIPTION
**Pull request by vasa-develop**
- [x] Adding AvionDB specific types (user facing functions)
- [x] Adding AvionDB specific types (internal functions)
- [x] Adding OrbitDB specific types
- [x] Adding IPFS specific types

License: MIT
Signed-off-by: Vaibhav Saini <vasa@dappkit.io>

### CODE REVIEW

File: src/core/Store.ts [hunk 1]
```diff
@@ -1,18 +1,19 @@
...
   _type: string;
-  _orbitdb: any;
-  openCollections: {};
+  _orbitdb: OrbitDB;
+  openCollections: any;
   _index: any;
   events: any;
   address: any;
-  constructor(ipfs, id, dbname: string, options: any = {}) {
+  constructor(ipfs: any, id: any, dbname: string, options: any = {}) {
```
Comment (deanpress) [inline src/core/Store.ts:16]: `ipfs` should be of type `IPFS` from the js-ipfs package right?

File: src/core/Store.ts [hunk 2]
```diff
@@ -25,26 +26,36 @@ class Store extends OrbitdbStore {
 
     this._index.store = this;
 
-    this.events.on("write", (address, entry) => {
-      this._index.handleEntry(entry);
-    });
-    this.events.on("replicate.progress", (address, hash, entry) => {
+    this.events.on("write", (address: any, entry: any) => {
```
Comment (deanpress) [inline src/core/Store.ts:29]: Why are `address` and `entry` of `any` type?

File: src/core/index.ts [hunk 3]
```diff
@@ -1,23 +1,37 @@
...
 const { Key } = require("interface-datastore");
 let datastore = EnvironmentAdapter.datastore(EnvironmentAdapter.repoPath());
 
 class AvionDB extends Store {
   static Collection = Collection;
 
-  static async init(name, ipfs, options: any = {}, orbitDbOptions) {
+  static async init(
+    name: string,
+    ipfs: any,
```
Comment (deanpress) [inline src/core/index.ts:17]: Same here - I think you can confidently make `ipfs`  of the IPFS type

File: src/core/Store.ts (hunk 1 above)
Comment (vasa-develop) [reply src/core/Store.ts:16]: Yes, but as we don't import the IPFS class within the `aviondb` module, I can't pass the class directly. We can use IPFS Type definitions (I'll have to look for them). Any other ideas on how we can add the IPFS type?

File: src/core/Store.ts (hunk 2 above)
Comment (vasa-develop) [reply src/core/Store.ts:29]: I wasn't sure about these. I'll log and see what can the possible types. Maybe @vaultec81 can tell us.

File: src/core/index.ts (hunk 3 above)
Comment (vasa-develop) [reply src/core/index.ts:17]: Same as above. How should I add IPFS type without importing IPFS in the file?

File: src/core/Store.ts (hunk 1 above)
Comment (vasa-develop) [reply src/core/Store.ts:16]: @deanpress This seems to be a good way to add `IPFS` type: https://github.com/ceramicnetwork/js-ceramic/blob/9c805a12951b89cdf6f420788543c462e962c4f9/packages/3id-did-resolver/src/%40types/ipfs.d.ts

Comment (deanpress) [reply src/core/Store.ts:16]: @vasa-develop Agree - we can just include the interface. There's another one here with js-ipfs types that was made last month: https://github.com/Xmader/ipfs-types

File: src/core/Store.ts (hunk 2 above)
Comment (vaultec81) [reply src/core/Store.ts:29]: Address should be type of string. Entry should be type of object.

**Comment by onichandame** [comment]
mind sharing how to access the ipfs types in projects that have aviondb installed as a dependency?

**Pull request merged**